src/DependencyInjection/Compiler/BusBuilder/BusBuilder.php

```diff
@@ -0,0 +1,126 @@
+        $this->middlewareIds = $middlewareIds;
+    }
+
+    public function id(): string
+    {
+        return $this->busId;
+    }
+
+    public function serviceId(): string
+    {
+        return "tactician.commandbus.$this->busId";
+    }
+
+    public function locatorServiceId()
+    {
+        return "tactician.commandbus.{$this->busId}.handler.locator";
+    }
+
+    public function commandHandlerMiddlewareId(): string
+    {
+        return "tactician.commandbus.{$this->busId}.middleware.command_handler";
+    }
+
+    public function registerInContainer(ContainerBuilder $container, array $commandsToAccept)
+    {
+        $this->registerLocatorService($container, $commandsToAccept);
+
+        $container->setDefinition(
+            $this->commandHandlerMiddlewareId(),
+            new Definition(
+                CommandHandlerMiddleware::class,
+                [
+                    new Reference('tactician.handler.command_name_extractor.class_name'),
+                    new Reference($this->locatorServiceId()),
+                    new Reference($this->methodInflectorId),
+                ]
+            )
+        );
+
+        $container->setDefinition(
+            $this->serviceId(),
+            new Definition(
+                CommandBus::class,
+                [
+                    array_map(
+                        function (string $id) { return new Reference($id); },
+                        $this->middlewareIds
+                    )
+                ]
+            )
+        );
+    }
+
+    private function registerLocatorService(ContainerBuilder $container, $commandsToAccept)
+    {
+        // Leverage symfony/dependency-injection:^3.3 service locators
+        if (class_exists(ServiceLocator::class)) {
+            $definition = new Definition(
+                ContainerLocator::class,
+                [new Reference($this->registerHandlerServiceLocator($container, $commandsToAccept)), $commandsToAccept]
+            );
+        } else {
+            $definition = new Definition(
+                ContainerBasedHandlerLocator::class,
+                [new Reference('service_container'), $commandsToAccept]
+            );
+        }
+
+        $container->setDefinition($this->locatorServiceId(), $definition);
+    }
+
+    private function registerHandlerServiceLocator(ContainerBuilder $container, array $commandsToAccept): string
+    {
+        $handlers = [];
+        foreach ($commandsToAccept as $commandName => $id) {
+            $handlers[$id] = new ServiceClosureArgument(new Reference($id));
+        }
+
+        $handlerServiceLocator = (new Definition(ServiceLocator::class, [$handlers]))
+            ->setPublic(false)
+            ->addTag('container.service_locator');
+
+        $container->setDefinition(
+            $id = "tactician.commandbus.{$this->busId}.handler.service_locator",
+            $handlerServiceLocator
+        );
+
+        return $id;
+    }
+}
```

Comment:
Several files are missing a trailing new line, not sure if that's an issue, but I figured I'd mention it anyway

Reply:
Drat, I recently nuked my PHPStorm config. Will fix this and probably run CS-Fixer on the whole thing at some point.